IBM-HPC: does not build on Power8 with GCC due to vector unit checking and instrinsics

We are trying to build hh-suite on Power8.  Our configuration looks like:

```
git clone https://github.com/soedinglab/hh-suite.git
cd hh-suite
git submodule init
git submodule update
mkdir build && cd build

#requires recent cmake 3.5.1 was used
cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX=/opt/um-ppc64le/hh-suite ..
```

Appears from the error message that cmake checks for SSSE3 and then the code uses abstracted versions of different vector calls in `src/simd.h`.

We would like to add support for Power 8  VSX when compiling on `ppc64le`.  To do this we are assuming

`cmake/CheckSSEFeatures.cmake`
`src/simd.h`

Both need logic added for the VSX intrinsics.  In the end we want a cmake and build that works on Power8.

```
$ cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX=/opt/um-ppc64le/hh-suite ..
-- The CXX compiler identification is GNU 4.8.3
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Compiler is GNU 
-- Using CPU native flags for SSE optimization:  -march=native
-- Performing Test HAVE_MM_MALLOC
-- Performing Test HAVE_MM_MALLOC - Failed
-- Performing Test HAVE_POSIX_MEMALIGN
-- Performing Test HAVE_POSIX_MEMALIGN - Success
-- Performing Test HAVE_AVX2_EXTENSIONS
-- Performing Test HAVE_AVX2_EXTENSIONS - Failed
-- Performing Test HAVE_AVX_EXTENSIONS
-- Performing Test HAVE_AVX_EXTENSIONS - Failed
-- Performing Test HAVE_SSE4_2_EXTENSIONS
-- Performing Test HAVE_SSE4_2_EXTENSIONS - Failed
-- Performing Test HAVE_SSE4_1_EXTENSIONS
-- Performing Test HAVE_SSE4_1_EXTENSIONS - Failed
-- Performing Test HAVE_SSSE3_EXTENSIONS
-- Performing Test HAVE_SSSE3_EXTENSIONS - Failed
-- Performing Test HAVE_SSE3_EXTENSIONS
-- Performing Test HAVE_SSE3_EXTENSIONS - Failed
-- Performing Test HAVE_SSE2_EXTENSIONS
-- Performing Test HAVE_SSE2_EXTENSIONS - Failed
-- Performing Test HAVE_SSE_EXTENSIONS
-- Performing Test HAVE_SSE_EXTENSIONS - Failed
-- No SSE extensions found
CMake Error at src/CMakeLists.txt:26 (message):
  SSSE3 is needed to run compile! CMake will exit.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,6 +19,25 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
   set(CMAKE_COMPILER_IS_ICC 1)
 endif()
 
+
+message("-- Processor is ${CMAKE_SYSTEM_PROCESSOR}")
+
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64.*|x86_64.*|AMD64.*")
+  set(X86_64 1)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i686.*|i386.*|x86.*|amd64.*|AMD64.*")
+  set(X86 1)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm.*|ARM.*)")
+  set(ARM 1)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64.*|AARCH64.*)")
+  set(AARCH64 1)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "PPC64*|ppc64*|powerpc64*")
+  set(PPC64 1)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "PPC*|ppc*|powerpc*")
+  set(PPC 1)
+endif()
+
+
+
 set (HHSUITE_VERSION_MAJOR 3)
 set (HHSUITE_VERSION_MINOR 0)
 set (HHSUITE_VERSION_PATCH 0)

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -8,6 +8,9 @@ if(${HAVE_SANITIZER})
 	include(FindTSan)
 endif()
 
+
+if(X86 OR X86_64)
+
 #SSE
 if(${HAVE_AVX2})
     ADD_DEFINITIONS("-DAVX2")
@@ -26,6 +29,12 @@ else()
         message( FATAL_ERROR "SSSE3 is needed to run compile! CMake will exit." )
     endif(${HAVE_AVX2_EXTENSIONS})
 endif()
+
+else() #
+
+message("!! NOT X86 OR X86_64 !!")
+
+endif() # NOT X86 OR X86_64
 
 
 

--- a/src/hhblits.h
+++ b/src/hhblits.h
@@ -28,7 +28,10 @@
 #ifdef OPENMP
 #include <omp.h>
 #endif
+
+#ifdef SSE
 #include <emmintrin.h>
+#endif
 
 using std::cout;
 using std::cerr;

--- a/src/hhhmmsimd.h
+++ b/src/hhhmmsimd.h
@@ -5,8 +5,9 @@
 #include "hhhmm.h"
 #include "simd.h"
 
+#ifdef SSE
 #include <emmintrin.h>
-
+#endif
 
 class HMMSimd
 {

--- a/src/hhviterbi.h
+++ b/src/hhviterbi.h
@@ -122,6 +122,7 @@ class Viterbi {
         int elem, bool self, int par_min_overlap);
 
 
+#if defined(SSE) // X86* arch only for _mm_prefetch 
     static inline simd_float ScalarProd20Vec(simd_float* qi, simd_float* tj) {
       _mm_prefetch((char * ) &qi[4], _MM_HINT_T0);
       _mm_prefetch((char * ) &tj[4], _MM_HINT_T0);
@@ -158,6 +159,35 @@ class Viterbi {
       return simdf32_add(res0, res2);
 
     }
+#else
+    static inline simd_float ScalarProd20Vec(simd_float* qi, simd_float* tj) {
+
+      simd_float res0 = simdf32_mul(tj[0], qi[0]);
+      simd_float res1 = simdf32_mul(tj[1], qi[1]);
+      simd_float res2 = simdf32_mul(tj[2], qi[2]);
+      simd_float res3 = simdf32_mul(tj[3], qi[3]);
+      res0 = simdf32_add(simdf32_mul(tj[ 4],qi[ 4]), res0);
+      res1 = simdf32_add(simdf32_mul(tj[ 5],qi[ 5]), res1);
+      res2 = simdf32_add(simdf32_mul(tj[ 6],qi[ 6]), res2);
+      res3 = simdf32_add(simdf32_mul(tj[ 7],qi[ 7]), res3);
+      res0 = simdf32_add(simdf32_mul(tj[ 8],qi[ 8]), res0);
+      res1 = simdf32_add(simdf32_mul(tj[ 9],qi[ 9]), res1);
+      res2 = simdf32_add(simdf32_mul(tj[10],qi[10]), res2);
+      res3 = simdf32_add(simdf32_mul(tj[11],qi[11]), res3);
+      res0 = simdf32_add(simdf32_mul(tj[12],qi[12]), res0);
+      res1 = simdf32_add(simdf32_mul(tj[13],qi[13]), res1);
+      res2 = simdf32_add(simdf32_mul(tj[14],qi[14]), res2);
+      res3 = simdf32_add(simdf32_mul(tj[15],qi[15]), res3);
+      res0 = simdf32_add(simdf32_mul(tj[16],qi[16]), res0);
+      res1 = simdf32_add(simdf32_mul(tj[17],qi[17]), res1);
+      res2 = simdf32_add(simdf32_mul(tj[18],qi[18]), res2);
+      res3 = simdf32_add(simdf32_mul(tj[19],qi[19]), res3);
+      res0 = simdf32_add(res0, res1);
+      res2 = simdf32_add(res2, res3);
+      return simdf32_add(res0, res2);
+
+    }
+#endif
 
     // Calculate secondary structure score between columns i and j of two HMMs (query and template)
     static inline float ScoreSS(const HMM* q, const HMM* t, const int i,

--- a/src/hhviterbialgorithm.cpp
+++ b/src/hhviterbialgorithm.cpp
@@ -104,7 +104,7 @@ void Viterbi::AlignWithOutCellOff(HMMSimd* q, HMMSimd* t,ViterbiMatrix * viterbi
                                                            -1, -1, -1,  -1,  0,  4,  8, 12, -1, -1, -1, -1, -1, -1, -1, -1);
 #endif
 #ifdef VITERBI_CELLOFF
-    const __m128i tmp_vec        = _mm_set_epi32(0x40000000,0x00400000,0x00004000,0x00000040);//01000000010000000100000001000000
+    const simd_int tmp_vec = simdi32_set4(0x40000000,0x00400000,0x00004000,0x00000040); // _mm_set_epi32() // 01000000010000000100000001000000
 #ifdef AVX2
     const simd_int co_vec               = _mm256_inserti128_si256(_mm256_castsi128_si256(tmp_vec), tmp_vec, 1);
     const simd_int float_min_vec     = (simd_int) _mm256_set1_ps(-FLT_MAX);
@@ -408,7 +408,8 @@ void Viterbi::AlignWithOutCellOff(HMMSimd* q, HMMSimd* t,ViterbiMatrix * viterbi
             /* efgh                                            0000  0000  HGFE  0000 */
             const __m128i efgh     = _mm256_extracti128_si256(abcdefgh, 1);
             _mm_storel_epi64((__m128i*)&sCO_MI_DG_IM_GD_MM_vec[j], _mm_or_si128(abcd, efgh));
-#else
+#elif defined(SSE)
+
             byte_result_vec = _mm_packs_epi32(byte_result_vec, byte_result_vec);
             byte_result_vec = _mm_packus_epi16(byte_result_vec, byte_result_vec);
             int int_result  = _mm_cvtsi128_si32(byte_result_vec);

--- a/src/hhviterbialgorithm.cpp
+++ b/src/hhviterbialgorithm.cpp
@@ -104,8 +104,8 @@ void Viterbi::AlignWithOutCellOff(HMMSimd* q, HMMSimd* t,ViterbiMatrix * viterbi
                                                            -1, -1, -1,  -1,  0,  4,  8, 12, -1, -1, -1, -1, -1, -1, -1, -1);
 #endif
 #ifdef VITERBI_CELLOFF
-    const simd_int tmp_vec = simdi32_set4(0x40000000,0x00400000,0x00004000,0x00000040); // _mm_set_epi32() // 01000000010000000100000001000000
 #ifdef AVX2
+    const __m128i tmp_vec = _mm_set_epi32(0x40000000,0x00400000,0x00004000,0x00000040);//01000000010000000100000001000000
     const simd_int co_vec               = _mm256_inserti128_si256(_mm256_castsi128_si256(tmp_vec), tmp_vec, 1);
     const simd_int float_min_vec     = (simd_int) _mm256_set1_ps(-FLT_MAX);
     const simd_int shuffle_mask_celloff = _mm256_set_epi8(
@@ -118,6 +118,7 @@ void Viterbi::AlignWithOutCellOff(HMMSimd* q, HMMSimd* t,ViterbiMatrix * viterbi
                                                           3, 2,  1, 0,
                                                           3, 2,  1, 0);
 #else // SSE case
+    const simd_int tmp_vec = simdi32_set4(0x40000000,0x00400000,0x00004000,0x00000040);
     const simd_int co_vec = tmp_vec;
     const simd_int float_min_vec = (simd_int) simdf32_set(-FLT_MAX);
 #endif

--- a/src/simd.h
+++ b/src/simd.h
@@ -321,8 +321,154 @@ typedef __m128i simd_int;
 #define simdi32_srli(x,y)	_mm_srli_epi32(x,y) // shift integers in a right by y
 #define simdi32_i2f(x) 	    _mm_cvtepi32_ps(x)  // convert integer to s.p. float
 #define simdi_i2fcast(x)    _mm_castsi128_ps(x)
+
+#define simdi32_set4(x,y,z,t) _mm_set_epi32(x,y,z,t)  // Added with Power8, hhviterbialgorithm needs _set4
+
 #endif //SIMD_INT
 #endif //SSE
+
+
+/*
+ *  Power8/LE Altivec/VSX SIMD
+ */
+#ifdef __ALTIVEC__ // || __VSX__
+
+#include <altivec.h>
+
+// double support
+#ifndef SIMD_DOUBLE
+#define SIMD_DOUBLE
+#define ALIGN_DOUBLE    16
+#define VECSIZE_DOUBLE  2
+typedef __vector double simd_double;
+#define simdf64_add(x,y)    vec_add(x,y)
+#define simdf64_sub(x,y)    vec_sub(x,y)
+#define simdf64_mul(x,y)    vec_mul(x,y)
+#define simdf64_div(x,y)    vec_div(x,y)
+#define simdf64_max(x,y)    vec_max(x,y)
+#define simdf64_load(x)     vec_vsx_ld(0,x)   // vec_ld
+#define simdf64_store(x,y)  vec_vsx_st(y,0,x) // vec_st
+#define simdf64_set(x)      vec_splats(x)
+//#define simdf64_set2(x,y)   _mm_set_ps(x,y)
+#define simdf64_setzero(x)  vec_splats(0)
+#define simdf64_gt(x,y)     vec_cmpgt(x,y)
+#define simdf64_lt(x,y)     vec_cmplt(x,y)
+#define simdf64_or(x,y)     vec_or(x,y)
+#define simdf64_and(x,y)    vec_and(x,y)
+#define simdf64_andnot(x,y) vec_nand(x,y)
+#define simdf64_xor(x,y)    vec_xor(x,y)
+#endif // SIMD_DOUBLE
+
+// float support
+#ifndef SIMD_FLOAT
+#define SIMD_FLOAT
+#define ALIGN_FLOAT     16
+#define VECSIZE_FLOAT   4
+typedef __vector float simd_float;
+#define simdf32_add(x,y)    x + y  //vec_add(x,y)
+#define simdf32_sub(x,y)    x - y  //vec_sub(x,y)
+#define simdf32_mul(x,y)    x * y  //vec_mul(x,y)
+#define simdf32_div(x,y)    x / y  //vec_div(x,y)
+#define simdf32_max(x,y)    vec_max(x,y)
+#define simdf32_min(x,y)    vec_min(x,y)
+#define simdf32_rcp(x)      vec_re(x)
+#define simdf32_load(x)     vec_vsx_ld(0,x)   // vec_ld
+#define simdf32_store(x,y)  vec_vsx_st(y,0,x) // vec_st
+#define simdf32_set(x)      vec_splats((float)x)
+//#define simdf32_set2(x,y)   _mm_set_ps(x,y)
+//#define simdf32_set4(x,y,z,t)
+#define simdf32_setzero(x)  vec_splats(0)
+#define simdf32_gt(x,y)     (simd_float)vec_cmpgt(x,y)
+#define simdf32_eq(x,y)     (simd_float)vec_cmpeq(x,y)
+#define simdf32_lt(x,y)     (simd_float)vec_cmplt(x,y)
+#define simdf32_or(x,y)     vec_or(x,y)
+#define simdf32_and(x,y)    vec_and(x,y)
+#define simdf32_andnot(x,y) vec_nand(x,y)
+#define simdf32_xor(x,y)    vec_xor(x,y)
+#define simdf32_extract(x,imm) vec_extract(x,imm)
+
+#define simdf32_f2i(x)      vec_cts(x,0)  // convert s.p. float to integer
+#define simdf_f2icast(x)    (simd_int)(x) // compile time cast
+#endif // SIMD_FLOAT
+
+// integer support 
+#ifndef SIMD_INT
+#define SIMD_INT
+#define ALIGN_INT       16
+#define VECSIZE_INT     4
+typedef __vector int simd_int;
+typedef __vector   signed char simd_s8;
+typedef __vector unsigned char simd_u8;
+
+#define simdi32_add(x,y)    vec_add(x,y)
+#define simdi32_sub(x,y)    vec_sub(x,y)
+#define simdi32_mul(x,y)    vec_mul(x,y)
+#define simdi32_max(x,y)    vec_max(x,y)
+#define simdi_load(x)       vec_vsx_ld(0,x)
+#define simdi_store(x,y)    vec_vsx_st(y,0,x)
+#define simdi32_set(x)      vec_splats((signed int)x)
+#define simdi32_set4(x,y,z,t) (simd_int){x,y,z,t}
+#define simdi_setzero(x)    vec_splats(0)
+#define simdi32_gt(x,y)     (simd_int)vec_cmpgt(x,y)
+#define simdi32_lt(x,y)     (simd_int)vec_cmplt(x,y)
+#define simdi_or(x,y)       vec_or(x,y)
+#define simdi_and(x,y)      vec_and(x,y)
+#define simdi_andnot(x,y)   vec_nand(x,y)
+#define simdi_xor(x,y)      vec_xor(x,y)
+#define simdi32_slli(x,y)   vec_sll(x,vec_splats((unsigned)y)) // shift integers in a left by y
+#define simdi32_srli(x,y)   vec_srl(x,vec_splats((unsigned)y)) // shift integers in a right by y
+#define simdi32_i2f(x)      vec_ctf(x,0)  // convert integer to s.p. float
+#define simdi_i2fcast(x)    (simd_float)(x)
+#define simdi8_set(x)       (simd_int)vec_splats((unsigned char)x)
+#define simdi8_gt(x,y)      (simd_int)vec_cmpgt((simd_s8)x,(simd_s8)y)
+#define simdi8_eq(x,y)      (simd_int)vec_cmpeq((simd_s8)x,(simd_s8)y)
+#define simdui8_max(x,y)    (simd_int)vec_max((vector unsigned char)x, (vector unsigned char)y)
+#define simdui8_adds(x,y)   (simd_int)vec_adds((simd_u8)x,(simd_u8)y)
+#define simdui8_subs(x,y)   (simd_int)vec_subs((simd_u8)x,(simd_u8)y)
+#define simdi8_shiftl(x,y)   (simd_int)vec_sll(x,vec_splats((char)y)) // shift integers in a left by y
+#define simdi8_shiftr(x,y)   (simd_int)vec_srl(x,vec_splats((char)y)) // shift integers in a right by y
+#define simdi8_movemask(x)  v_movemask(x)
+
+
+// There is no altivec/vsx equivalent, C version 
+//
+inline int v_movemask(simd_int x)
+{
+  unsigned int result=0;
+
+  union {
+    simd_int si;
+    char as_char[16];
+  } t;
+
+  t.si = x;
+  result |= (t.as_char[15] & 0x80) << (15-7);    
+  result |= (t.as_char[14] & 0x80) << (14-7);   
+  result |= (t.as_char[13] & 0x80) << (13-7);   
+  result |= (t.as_char[12] & 0x80) << (12-7);   
+  result |= (t.as_char[11] & 0x80) << (11-7);   
+  result |= (t.as_char[10] & 0x80) << (10-7);   
+  result |= (t.as_char[9]  & 0x80) <<  (9-7);   
+  result |= (t.as_char[8]  & 0x80) <<  (8-7);   
+  result |= (t.as_char[7]  & 0x80);   
+  result |= (t.as_char[6]  & 0x80) >>  (7-6);   
+  result |= (t.as_char[5]  & 0x80) >>  (7-5);   
+  result |= (t.as_char[4]  & 0x80) >>  (7-4);   
+  result |= (t.as_char[3]  & 0x80) >>  (7-3);   
+  result |= (t.as_char[2]  & 0x80) >>  (7-2);   
+  result |= (t.as_char[1]  & 0x80) >>  (7-1);   
+  result |= (t.as_char[0]  & 0x80) >>   7;    
+
+  return result;
+}
+
+
+#endif // SIMD_INT
+
+#endif // __ALTIVEC__ || __VSX__
+
+
+
 
 
 /* horizontal max */

